index.bs

```diff
@@ -432,10 +434,29 @@ the computation graph used to compute one or more specified outputs. The key
 purpose of the compilation step is to enable optimizations that span two or
 more operations, such as operation or loop fusion.
 
-The {{MLGraph/compute()}} method of the {{MLGraph}} interface is used to execute the
-compiled computation graph (to perform inference). The caller supplies the input
-values using {{MLNamedInputs}}, binding the input {{MLOperand}}s to their values.
-The caller supplies pre-allocated buffers for output {{MLOperand}}s using {{MLNamedOutputs}}.
+Once the {{MLGraph}} is constructed, there are multiple ways by which the graph may be executed. The
+{{MLContext/compute()}} method represents a way the execution of the graph is carried out immediately 
+on the calling thread, which must also be a worker thread. The execution produces the results of the computation 
+from all the inputs bound to the graph. This type of execution is limited only to when the computational device 
+bound to the context is a CPU device.
```

Comment:
@dontcallmedom The GPU execution is inherently async but not on the app's timeline. It's async on the GPU timeline. So having a sync call on a GPU device would mean stalling the CPU until the GPU queue has a chance to be executed. That is undesirable and none of our known use case will do that.

Reply:
The TF.js use case on a GPU device is such that the execution call is done on the main thread and not the worker thread. However, the compute result is deferred until the GPU executes the result and the readback occurs. The reason the sync API is limited to only the worker thread is for a different purpose. It's because even for the CPU execution, the main thread should never block. 

Again, this change assumes that the app knows what type of MLContext it creates. WebNN makes that assumption.

Reply:
> the app knows what type of MLContext it creates. WebNN makes that assumption.

To clarify my point - I'm not arguing here (as I did in #255) that the `MLContext` may be coming from a third party library. My assertion is that as a developer, even if I set `MLDdevicePreference` to CPU, the spec gives no guarantee that I'll get back a CPU-based `MLContext`, and doesn't even expose if the `MLContext` I got back is CPU-based or not. So as a developer, I don't know if I can or cannot execute the sync version of the API, except by actually executing it and getting an exception (which I don't think is a practical API to use).

Reply:
> the MLContext may be coming from a third party library

I think we will need to establish if this is a supported use case. As of the current design goal, since WebNN is designed to be a framework backend, it is highly unlikely that the MLContext will be coming from a third party library. Frameworks such as TF.js owns the creation of MLContext and the execution that follows. 

>  if I set MLDdevicePreference to CPU, the spec gives no guarantee that I'll get back a CPU-based MLContext, and doesn't even expose if the MLContext I got back is CPU-based or not.

This is a separate issue and a valid concern. In order to support a default setting, there needs to be a defined default value for `MLDevicePreference` and that it's currently defined as a CPU device. To make it clearer, we could simply remove the `cpu` option from the enum. Should we go ahead and do that?

Reply:
> there needs to be a defined default value for MLDevicePreference and that it's currently defined as a CPU device. To make it clearer, we could simply remove the cpu option from the enum. Should we go ahead and do that?

I don't think this helps - the real question is whether there is a guarantee that requesting a default (or a CPU) device gives you a CPU MLContext or not. My reading of the spec is that there isn't such a guarantee.

Reply:
My point is to make that guarantee thru the `default` option, and eliminate the `cpu` option as it would be redundant.

Reply:
let's bring this to a separate issue - I think I would keep `"cpu"` as a name and instead make it the WebIDL-default value of `MLContextOptions.devicePreference`; but my sense is that the primary question around such a change is more about the impact of making `devicePreference` no longer a hint but a setting.

Reply:
We can use issue https://github.com/webmachinelearning/webnn/issues/169 for device selection discussion.

Reply:
I agree with making "cpu" a default option and remove "default" as @dontcallmedom suggested here. I don't think the word "preference" necessarily implies a hint.

Reply:
As I [commented](https://github.com/webmachinelearning/webnn/pull/255#discussion_r816500166) in #255, I think it should support GPU as well. It would allow the caller who requires sync API could also access GPU, e.g., TFLite WebNN delegate. When the compute device is GPU, it would block the calling thread until the result is readback into the output ArrayBuffers. This would not impact the responsiveness of the main (UI) thread as its usage is restricted in the DedicatedWorker.

Reply:
I'll defer this to @RafaelCintron who insisted that GPU sync call is bad for user experience.

Reply:
> let's bring this to a separate issue - I think I would keep `"cpu"` as a name and instead make it the WebIDL-default value of `MLContextOptions.devicePreference`; but my sense is that the primary question around such a change is more about the impact of making `devicePreference` no longer a hint but a setting.

In the latest commit, I've removed the `default` option and rename `MLDevicePreference` to `MLDeviceType` to make it more clear that it is not merely a hint. @dontcallmedom Please take a look.

Reply:
The way you explain it here is not really "default" but rather "auto" option i.e. automatic device selection. This is a hard subject with potential backward and forward compatibility guarantee involved. Also the definition of "best" varies greatly depending on what the application is trying to do. Generally speaking, this type of policy behavior should not belong to the backend component like the WebNN API. Rather, it should be part of the application logic or in the framework on behalf of the application. 

"default" and "auto" aren't really the same thing. Default is a well-defined behavior that is chosen to be a default behavior when nothing more explicit is specified, while the behavior of `auto` is dynamic implying the notion of "best fit", a more open-ended policy. 

I prefer that we discuss automatic device selection as a separate issue.